coderd/coderdtest/oidctest/idp.go

```diff
@@ -58,15 +59,107 @@ type deviceFlow struct {
 	granted   bool
 }
 
+type fakeIDPProtected struct {
+	mu sync.RWMutex
+
+	issuer     string
+	issuerURL  *url.URL
+	key        *rsa.PrivateKey
+	provider   ProviderJSON
+	handler    http.Handler
+	cfg        *oauth2.Config
+	fakeCoderd func(req *http.Request) (*http.Response, error)
+}
+
+func (f *fakeIDPProtected) Issuer() string {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.issuer
+}
+
+func (f *fakeIDPProtected) IssuerURL() *url.URL {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.issuerURL
+}
+
+func (f *fakeIDPProtected) PrivateKey() *rsa.PrivateKey {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.key
+}
+
+func (f *fakeIDPProtected) Provider() ProviderJSON {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.provider
+}
+
+func (f *fakeIDPProtected) Config() *oauth2.Config {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.cfg
+}
+
+func (f *fakeIDPProtected) Handler() http.Handler {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.handler
+}
+
+func (f *fakeIDPProtected) SetIssuer(issuer string) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.issuer = issuer
+}
+
+func (f *fakeIDPProtected) SetIssuerURL(issuerURL *url.URL) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.issuerURL = issuerURL
+}
+
+func (f *fakeIDPProtected) SetProvider(provider ProviderJSON) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.provider = provider
+}
+
+func (f *fakeIDPProtected) MutateConfig(fn func(cfg *oauth2.Config)) {
+	f.mu.RLock()
+	cfg := f.cfg
+	if cfg == nil {
+		cfg = &oauth2.Config{}
+	}
+	fn(cfg)
+	f.mu.RUnlock()
+	f.mu.Lock()
+	f.cfg = cfg
+	f.mu.Unlock()
+}
+
+func (f *fakeIDPProtected) SetHandler(handler http.Handler) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.handler = handler
+}
+
+func (f *fakeIDPProtected) SetFakeCoderd(fakeCoderd func(req *http.Request) (*http.Response, error)) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.fakeCoderd = fakeCoderd
+}
+
+func (f *fakeIDPProtected) FakeCoderd() func(req *http.Request) (*http.Response, error) {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.fakeCoderd
+}
+
 // FakeIDP is a functional OIDC provider.
 // It only supports 1 OIDC client.
 type FakeIDP struct {
-	issuer    string
-	issuerURL *url.URL
-	key       *rsa.PrivateKey
-	provider  ProviderJSON
-	handler   http.Handler
-	cfg       *oauth2.Config
+	prot fakeIDPProtected
```

Comment:
Suggestion: Maybe rename to priv/private, internal, etc? Prot makes me think of protoc/prototypes.

Reply:
Renamed to `locked`. 👍